Ticket Probability Fixes

mikefullman opened the pull request:
Proposed fixes for #281 

the ROI calculation for the main page was adjusted to take into consideration 50% likelihood days based on the following Poisson calculation of probabilities:

> Chance for a ticket to be voted in a given time frame:
> 1 - (1 - P)^N
> 
> Where:
> P is the probability of a vote in one block.
> N is the number of blocks before ticket expiry.

**Main Page**
ROI_DAYS= [[ ln(50%) / ln(1-(5/ACTUAL_TICKET_POOL)) ] + TICKET_MATURITY +COINBASE_MATURITY ] / BLOCKS_PER_DAY

**Transaction Page**
Added probability bar chart for this ticket based on time to expiry.  I'm not sure if this is what @raedah was referring to in https://github.com/decred/dcrdata/issues/281#issuecomment-374038922 or not but I felt it would be useful on this page.

Please let me know if you have any thoughts on the luck section.

**Additional Issues Noted**
The expiry time is incorrect.  Expiry starts at the end of the ticket maturity and therefore actual expiry is 40960+256 blocks after ticket purchase.  I would like to adjust that chart so its only visible for mature tickets and starts at block 0 right when the ticket matures, and 100% is at 40960.

**Screenshots**
![image](https://user-images.githubusercontent.com/11194546/39030952-3bb61cc2-441a-11e8-950c-fdf3902328c9.png)

![image](https://user-images.githubusercontent.com/11194546/39030967-5339126e-441a-11e8-8156-c9d4210f877f.png)

![image](https://user-images.githubusercontent.com/11194546/39030977-6205b43c-441a-11e8-95bc-5cc316f22956.png)

**Todo**

1. Fix Expiry on transactions page
2. Confirm boundary cases display the proper info for probability and expiry bar charts: immaturity -> maturity -> expiry.  Voted, revoked, expired, missed.
3. Check for divide by zero problems or at least see how Go handles them

chappjc commented:
Should the ROI period be the average vote time or the time where a ticket has a 50% chance of voting?  Say you have a large number of tickets.  It's the former.

chappjc commented:
I approve the expiry fix.  Never noticed that.  :)

The probability issue @raedah was referring to was from this: https://github.com/decred/dcrdata/pull/361#issuecomment-372167721.  So yes, it looks like you got that right.

chappjc commented:
Actually, I'd argue we want to compute ROI per month, since we're humans.  ;)  So I'd actually scale up the ROI that you get for the ~28 day period, and state "(monthly)".

But I still don't see a case for using the solution for p=0.5 for one ticket.  In aggregate, the average vote time, is relevant.

Consider computing the p at 30 days. It's not 0.5 because probability is not linearly related to time.  There is a slowly decaying tail out to 4.7 months, pulling the average up.

mikefullman commented:
As per offline dialog with developers
1.  Main page ROI was adjusted to use average + ticket maturity + coinbase maturity.
2.  Transaction pages were adjusted to have probability of NOT voting rather than probability of voting
3.  Expiry graph was fixed to go from 1 to 40960

The differences between existing beta explorer and this PR



| STATUS | OLD | NEW |
|---------|-------|-------|
| Just after buy | ![image](https://user-images.githubusercontent.com/11194546/39159715-f2934024-471b-11e8-8c0b-f4eab5fc8042.png) | ![image](https://user-images.githubusercontent.com/11194546/39159732-016e571e-471c-11e8-85d1-1e3e3fc99253.png) |
| Just before mature | ![image](https://user-images.githubusercontent.com/11194546/39159801-7256f800-471c-11e8-970b-a7d96b78e5b7.png) | ![image](https://user-images.githubusercontent.com/11194546/39159793-6bf44a76-471c-11e8-96b9-c606d22484a4.png) |
| First mature block | ![image](https://user-images.githubusercontent.com/11194546/39159886-0909ab8a-471d-11e8-9251-9b46dc78b0c7.png) | ![image](https://user-images.githubusercontent.com/11194546/39159877-f6a15466-471c-11e8-9e1d-bf7f0e278004.png) |
| Unvoted | ![image](https://user-images.githubusercontent.com/11194546/39160031-eca3bb6a-471d-11e8-9123-3aff45beac72.png)  | ![image](https://user-images.githubusercontent.com/11194546/39160009-c367b74c-471d-11e8-8da3-371c720f5b29.png) |
| Voted and Mature | ![image](https://user-images.githubusercontent.com/11194546/39159969-966a35a8-471d-11e8-8b88-e2d946403cdd.png) | ![image](https://user-images.githubusercontent.com/11194546/39159961-8235ac70-471d-11e8-9e18-138d08d3e204.png) | 
| Expired | ![image](https://user-images.githubusercontent.com/11194546/39160313-6fb17b36-471f-11e8-98bc-11dbb0260649.png) | ![image](https://user-images.githubusercontent.com/11194546/39160295-593df6b8-471f-11e8-9f6c-d3d1cf240b45.png) |

chappjc commented on explorer/explorerroutes.go:
```diff
@@ -241,7 +245,15 @@ func (exp *explorerUI) TxPage(w http.ResponseWriter, r *http.Request) {
+				// P is the probability of a vote in one block. (votes per block / current ticket pool size)
+				// N is the number of blocks before ticket expiry. (ticket expiry in blocks - (number of blocks since ticket purchase - ticket maturity))
+				// C is the probability (chance)
+				tx.TicketInfo.Probability = (math.Pow((1 - (float64(exp.ChainParams.TicketsPerBlock) / float64(exp.ExtraInfo.PoolInfo.Size))), (float64(exp.ChainParams.TicketExpiry) - (float64(tx.Confirmations) - float64(exp.ChainParams.TicketMaturity))))) * 100
```
By splitting this up, it would be easier to follow.

e.g.:
```
blocksLive := float64(tx.Confirmations) - float64(exp.ChainParams.TicketMaturity)
pVote := float64(exp.ChainParams.TicketsPerBlock) / float64(exp.ExtraInfo.PoolInfo.Size)
// or just pVote := 1 / exp.ChainParams.TicketPoolSize
...

chappjc commented on views/tx.tmpl:
```diff
@@ -53,110 +53,110 @@
+                    {{if and (and (eq .Mature "True")  (ne .TicketInfo.SpendStatus "Voted")) (eq .TicketInfo.PoolStatus "live")}}
+                        {{if ge .TicketInfo.TicketExpiry .TicketInfo.ShortConfirms}}
+                            <tr>
+                                <td class="text-right pr-2 h1rem p03rem0">PROBABILITY</td>
```
I'm starting to think this would be more accessible to the user phrased like "Chance of expiry"

chappjc commented:
Need to fix this actually:

    explorer/explorer.go:1::warning: file is not gofmted with -s (gofmt)

raedah commented:
Luck on Expired should say 'final' instead of 'possible'.

raedah commented:
LGTM. Maybe make the c in chance lower case.

mikefullman commented:
It didn't make sense to me at first either.  My first thought was to agree with you, but since we already have another var that is similar, _blocksLive_ which is close but not quite what we want with ShortConfirms, I would recommend we keep the variable name but update the definition comments a bit to make it easier and more precise to understand.  Happy to do whatever you want though.

old:
```go
ShortConfirms        int64   // Total number of confirms up until the point the ticket votes or expires
```

new:
```go
ShortConfirms        int64   // Total number of confirms after maturity up until the point the ticket votes or expires
```

mikefullman commented:
Great feedback!  Happy to look at that.  There are three outstanding issues now as far as I know.  I wasn't sure how much to keep pulling into this PR.

1.  I could go ahead and fix #440. (immature coinbase transactions do not show correct bar chart)
2.  I noted with @raedah that the existing design makes use of things like ``if tx.Type == "Ticket" {`` and also ``if tx.TicketInfo.SpendStatus == "Voted" {`` and in the template `` {{if and (and (eq .Mature "True")  (ne .TicketInfo.SpendStatus "Voted")) (eq .TicketInfo.PoolStatus "live")}}``.  These types of things can cause problems later when English words are used in comparisons to control program functionality.  If we decide to ever do a translation or even change one of those things from "live" to "Live" then it can cause bugs.  I would recommend using const's instead of english word comparisons for program flow and pull the English words in last second in the view.  @raedah recommended a separate PR.  I just haven't gotten to putting the issue in.
3.  Missed/revoked ticket rendering.  I recommend that I put in a change to remove the luck line item for missed tickets regardless of their revoked/unspent status.  The expiry chart is already not present.  The only other thing to consider is revoked ticket maturity status.  I need to explore if ticket funds are immediately available (as I see when I look at some transactions) or if they need 256 blocks of maturity as the docs specify.    https://docs.decred.org/faq/proof-of-stake/voting-tickets/#6-what-happens-if-a-ticket-i-have-missed-a-vote

Here is existing status of missed/revoked tickets:
(note, expired doesn't mean much for a missed ticket, but I included it below because the existing design renders an expiry chart)

| STATUS | OLD | NEW |
|---------|-------|-------|
| Missed / Unspent |![image](https://user-images.githubusercontent.com/11194546/39455172-fa22bcca-4c93-11e8-8b52-d2330c681174.png) | ![image](https://user-images.githubusercontent.com/11194546/39455465-532aecb0-4c95-11e8-8d1c-cb21d843a10f.png) |
| Missed / Revoked / _unexpired_ |![image](https://user-images.githubusercontent.com/11194546/39455197-14297bfe-4c94-11e8-9463-c08ab80b9818.png) | ![image](https://user-images.githubusercontent.com/11194546/39455526-bd4a8128-4c95-11e8-9053-028066c88e45.png) |
| Missed / Revoked / _expired_ |![image](https://user-images.githubusercontent.com/11194546/39455230-3e901b96-4c94-11e8-8f32-72c5d666cf22.png) | ![image](https://user-images.githubusercontent.com/11194546/39455320-a5862d22-4c94-11e8-9c6e-d61328705fd4.png) |

chappjc commented:
Regarding (2), the use of string literals, I absolutely agree we should address this, but it's not necessary for this PR.  However, I'd suggest an approach like I used here:

https://github.com/decred/dcrdata/blob/e1c4f606aeab37e6171ee2677f701b89970c4c69/db/dbtypes/types.go#L39-L104

For missed, I believe there is an argument to be made for still showing luck since the ticket was still called in a lottery, but (possibly) the owner of the ticket misconfigured something, unplugged their router or whatever.  So this is sort of a lucky but incompetent situation.  That said, I agree we could hide luck stuff since it has the potential to be confusing.  Being missed changes things.

Regarding maturity of revoked tickets, I'm pretty sure it does not apply since there is no subsidy payed by the revocation transaction.

chappjc commented:
BTW, thanks for checking all the missed / spent statuses and showing how they appear.  This is always very helpful for PRs.  I agree with removing the expiry bar in all three cases.

chappjc commented:
If the changes are small, I'd like to see the coinbase maturity (1) fixed in this PR.  It's fresh in your mind and you know what to do.

mikefullman commented:
Ok, I think we are good to go!
1.  I changed `ShortConfirms ` to `TicketLiveBlocks`
2.  I removed Luck from all missed tickets
3.  I fixed the coinbase transactions to properly reflect maturity after coinbasematurity blocks.
4.  I noticed the days until mature was not working for voted blocks - fixed that.
5.  I went back and spot checked previous test cases to make sure nothing changed.

**LUCK REMOVED**

| STATUS | OLD | NEW |
|---------|-------|-------|
| Missed / Unspent |![image](https://user-images.githubusercontent.com/11194546/39455172-fa22bcca-4c93-11e8-8b52-d2330c681174.png) | ![image](https://user-images.githubusercontent.com/11194546/39460341-8c63c980-4cb7-11e8-8ccf-1ab153fa55f6.png) |
| Missed / Revoked / _unexpired_ |![image](https://user-images.githubusercontent.com/11194546/39455197-14297bfe-4c94-11e8-9463-c08ab80b9818.png) | ![image](https://user-images.githubusercontent.com/11194546/39460358-b277925a-4cb7-11e8-8ee3-20ef7dcca543.png)  |
| Missed / Revoked / _expired_ |![image](https://user-images.githubusercontent.com/11194546/39455230-3e901b96-4c94-11e8-8f32-72c5d666cf22.png) | ![image](https://user-images.githubusercontent.com/11194546/39460395-ea9e7e00-4cb7-11e8-899c-68b676b7334e.png)  |

**COINBASE MATURITY**

| STATUS | OLD | NEW |
|---------|-------|-------|
|Vote, 3 confirmations | ![image](https://user-images.githubusercontent.com/11194546/39464038-1855c818-4cd0-11e8-95e4-fdb5e5556d65.png) | ![image](https://user-images.githubusercontent.com/11194546/39464079-511b894e-4cd0-11e8-924a-dae76d92e8bc.png) |
|coinbase, 255 confirms | ![image](https://user-images.githubusercontent.com/11194546/39463691-1af2a0b6-4cce-11e8-8edf-0a8dfd588709.png) | ![image](https://user-images.githubusercontent.com/11194546/39463699-2a6bd2c4-4cce-11e8-84cb-ef63996b4c77.png) |

chappjc commented:
tACK

mikefullman commented:
Great!

Sent from my iPhone

> On May 1, 2018, at 7:30 PM, Jonathan Chappelow <notifications@github.com> wrote:
> 
> Merged #436.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
